QUIC support

Demi-Marie opened the pull request:
I am using the latest quinn git master.  While I could backport to the latest release, I would prefer to get the existing code working first.

Current approach:

* Use `quinn-proto` (the bare state machine)
* Each connection has a pair of `HashMap`s to store wakers, one for readers and one for writers.  When I/O on a stream becomes possible, the corresponding waker is awoken.
* The mutable state is protected with mutexes, so `libp2p-quic` is thread-safe.
* The state machine is only advanced on a background task.
* Channels are used for incoming connections (to an endpoint) and incoming streams (to a connection).  They are also used to send packets generated by a connection to the background task for transmission.
* Strict mutex aquisition ordering is upheld to prevent deadlocks.  Specifically, if both the mutex protecting the endpoint and the mutex protecting a connection need to be taken, the mutex protecting the endpoint must be taken first.

What I would like feedback on:

* Should I be using channels in places where I am currently using mutexes?
* To avoid potential memory exhaustion denial of service attacks, I am using a bounded channel for outgoing packets.  If it fills up, packets are dropped.  Since QUIC is a reliable protocol, I expect that these will be retransmitted, which is much better than having to buffer an unlimited amount of data.
* How should I handle the case where a connection arrives, but we are not ready for it?  We can’t buffer an unlimited number of them, so at some point we will need to impose backpressure.  The current plan is to rely on quinn-proto limiting the number of connections that it returns but have not yet been accepted.

### Edit

The code is now essentially complete and is ready for review.

Demi-Marie committed:
Add libp2p-transport-quic crate

Demi-Marie committed:
Add copyright notice and beginnings of a transport

Demi-Marie committed:
Bogus implementation that at least compiles

Demi-Marie committed:
Remove an `unimplemented!`

Demi-Marie committed:
Report peer addresses

This requires a branch of `quinn` that has not yet been merged.

Demi-Marie committed:
Remove code duplication between QUIC and TCP transports

Demi-Marie committed:
Switch back to quinn master

It now incudes the changes needed by libp2p.

Demi-Marie committed:
Implement `dial`

Demi-Marie committed:
Add `use` lines for std future related types

Also update to stable futures.

Demi-Marie committed:
Initial StreamMuxer impl

There are a LOT of `unimplemented!()` calls here!

Demi-Marie committed:
Last version using ‘quinn’

All future versions will use ‘quinn-proto’ directly.

Demi-Marie committed:
Compiling (but not working) quinn-proto based libp2p-quic

There are plenty of `unimplemented!()` calls, and TLS certificate
processing has not even been started, but this should still be a good
start.

Demi-Marie committed:
More progress on libp2p-quic

The code compiles and is approaching a point at which it could be
useful.

Demi-Marie committed:
The doc test passes!

Demi-Marie committed:
Merge branch 'stable-futures' into demi-quic-stable-futures

Demi-Marie committed:
Preserve order of outgoing connections

Demi-Marie committed:
Test suite compiles!

Demi-Marie committed:
Handle making new connections

Demi-Marie committed:
Merge branch 'stable-futures' into demi-quic-stable-futures

Demi-Marie committed:
All tests compile

Demi-Marie committed:
Remove remaining unimplemented!()

Demi-Marie committed:
Merge branch 'stable-futures' into demi-quic-stable-futures

Demi-Marie committed:
Simple fixes

Demi-Marie committed:
Implement sending messages

This implements message transmission.  The code relies on UDP datagram
sending not blocking indefinitely.  Since UDP does not retransmit
packets, this should be a decent assumption in practice.

Demi-Marie committed:
Pin futures_codec to avoid compilation errors

Demi-Marie committed:
Fix the rest of the compiler errors

Demi-Marie committed:
Bump dependencies

This bumps a few dependencies to recent but working versions.

Demi-Marie committed:
Merge branch 'dm-bump-deps' into demi-quic-stable-futures

Demi-Marie committed:
Implement complete QUIC protocol logic

This implements the entire QUIC protocol logic (but no crypto).

Demi-Marie committed:
Fix warning in testsuite

We don’t actually need the the result of env_logger::try_init().  Just
drop it.

Demi-Marie committed:
Switch to parking_lot

Demi-Marie committed:
Outbound connections must fail when a connection is lost

Demi-Marie committed:
Add missing wakeups

The code would deadlock because certain wakeups were missed.

Demi-Marie committed:
Move all of the driver-waking code into a single function

Demi-Marie committed:
Consolodate the I/O code

Demi-Marie committed:
Refactor some common code

Demi-Marie committed:
More cleanups

Demi-Marie committed:
Formatting cleanups

Demi-Marie committed:
Implement certificate generation for libp2p

libp2p requrires custom certificate generation.  This implements it, but
does not hook it up to the rest of the codebase.

Demi-Marie committed:
Start working on certificate verification

Only the outermost level of parsing is implemented yet.  I will be doing
parsing manually, rather than using the x509-parser crate, as I don’t
need to extract all the information that crate extracts.

Demi-Marie committed:
Add connection.rs

This will someday hold connection code, once I do a refactor.

Demi-Marie committed:
Finish X.509 parsing

Extension verification is not yet implemented

Demi-Marie committed:
Some wrong attempts at certificate verification

This doesn’t pass its tests, but that is okay, as I will be swtiching to
a completely different approach: using webpki for certificate
verification.

Demi-Marie committed:
Use webpki to verify the certificate self-signature

This avoids massive X.509 headaches.

Demi-Marie committed:
Remove spurious debug assertion

Demi-Marie committed:
Remove commented-out code and excessive logging

Demi-Marie committed:
Cleanup certificate verification

Demi-Marie committed:
Remove connection.rs

It was going to be used for the connection struct, but there is not
enough code to justify this, at least for now.

Demi-Marie committed:
Get QUIC connections to work

This implements QUIC certificate generation.  Verification is not yet
implemented.

Demi-Marie committed:
Merge branch 'stable-futures' into demi-quic-stable-futures

Demi-Marie committed:
Fix the tests to avoid hangs

Demi-Marie committed:
testing

Demi-Marie committed:
Actually accept connections

This slows the tests down considerably.  I need to figure out why.

Demi-Marie committed:
Ensure the driver can actually be woken up

Demi-Marie committed:
Merge branch 'stable-futures' into demi-quic-stable-futures

Demi-Marie committed:
Futures need to be polled to completion!

Failure to do so caused missed wakeups and hangs.

Demi-Marie committed:
Clean up the driver and I/O code

Now the driver tasks take care of all I/O.

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Go much farther before hanging

There is still at least one missed wakeup.

Demi-Marie committed:
Make libp2p-quic more robust

There may be a rare, intermittent hang under certain conditions, but I
am having trouble reproducing it, and in libp2p it would be harmless ―
it would result in a timeout and immediate reconnection, which would
almost always succeed.  Libp2p needs to be able to handle such errors
anyway, since they can happen due to network problems, for example.
Therefore, I consider the futures portion of this code to be ready for
review.

Demi-Marie committed:
Cleanups

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Futures portion of libp2p-quic almost complete

The futures part of libp2p-quic is almost complete.  Only a small amount
of crypto code needs to be written.

Demi-Marie committed:
Move the endpoint code into a separate module

This makes the code clearer.

Demi-Marie committed:
Sending outgoing traffic must not block processing incoming traffic

This fixes a deadlock (that took weeks to debug) by processing received
packets even if there are packets waiting to be transmitted.  If packets
can’t be transmitted fast enough, they are simply dropped.  QUIC will
retransmit them, so that is okay.  We still log this at `warn` level,
since it is an undesirable occurrence.

The deadlock is the classic circular dependency that is found in
Ethernet networks using priority flow control.  UDP is an unreliable
protocol, but reliable, in-order delivery with flow control is still
allowed.  That is, there is no guarantee that the kernel will not block
senders for an arbitrary amount of time.  This is most likely to be an
issue on loopback, but could be an issue anywhere.

Also, increase the receive buffer size, to avoid problems on large MTU
networks.

There are still many problems that need to be fixed before this is
production-ready, including (but not limited to):

* Properly shutdown outgoint streams when `QuicMuxer::close()` is
  called.
* Do not loop excessively when incoming data arrives too quickly.
* Do not allocate a fresh receive buffer for each incoming packet.
* Properly cleanup background tasks.
* Add more tests.
* Report `PeerId`s for incoming connections.

Demi-Marie committed:
Another failed attempt to fix the hang!

This is almost certainly a but in a third-party library.

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Hopefully fix the hangs

Demi-Marie committed:
Fix the hangs for real

This version has, unless I made a huge mistake, run over 30,000 times in
a row without hanging.

Demi-Marie committed:
Fix wildcard_expansion test case

Demi-Marie committed:
Don’t use keep alives

They caused rare hangs that consumed 100% CPU.  This version has run for
over 528,000 successful runs.

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Add a keep-alive interval to prevent spurious timeouts

Demi-Marie committed:
Don’t use Poll::{self, Pending, Ready}

Demi-Marie committed:
Do not check for ECONNRESET on transmit

It can only happen on receive.

Demi-Marie committed:
Refactor networking code

Demi-Marie committed:
Cleanup I/O paths

This cleans up the I/O paths to unify logging and retry on ECONNRESET.
It also prepares for using a different mechanism to determine when an
endpoint can be shut down.

Demi-Marie committed:
Return peer IDs from Transport impl

This is necessary for use in higher-level protocols.

Demi-Marie committed:
Sending endpoint events could require checking for timers

Demi-Marie committed:
Move the connection code to its own module

This forced several other refactors, which resulted in a cleaner API.

Demi-Marie committed:
Don’t rely on a fork of `async-std`

The fork won’t be merged upstream, and `futures::pin_mut` makes it
unnecessary.

Demi-Marie committed:
Avoid a panic and work around a quinn-proto bug

It is possible to get multiple `drained` events, so the endpoint must
not panic in that case. Furthermore, due to
https://github.com/djc/quinn-proto/issues/604, quinn-proto can
spuriously return `Err(ReadError::Blocked)` instead of `Ok(None)` when
reading from a finished stream in a closed connection. Work around that.

Demi-Marie committed:
Remove unnecessary #[allow(dead_code)]

Demi-Marie committed:
Handle a critical Basic Constraints extension

We require certificates used in libp2p to not be certificate
authorities.

Demi-Marie committed:
Let WebPKI verify the certificate self-signature

This adds support for verifying notBefore and notAfter fields on the
certificate, and also removes a bunch of ugly code.

Demi-Marie committed:
Add error file

Demi-Marie committed:
Split up certificate verification

Part of it uses webpki and is now done from the rustls certificate
verifier.

Demi-Marie committed:
Merge branch 'master' into merged

Demi-Marie committed:
Simplify certificate verification

All X.509 certificates are validated by WebPKI before being parsed by
libp2p-quic.  As a result, libp2p-quic does not need to do even basic
validation itself.

Demi-Marie committed:
Clean up certificate verification code

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Don’t try to use `ifaddrs` in the browser

Demi-Marie committed:
Remove ugly workaround and clean up some code

https://github.com/djc/quinn/issues/617 has been fixed, which allows for
an ugly workaround (using the close reason to determine if a connection
close was clean) to be removed.

Demi-Marie committed:
Return an error if an unwritten stream is read

This avoids the deadlock that would otherwise result.

I also did some refactoring, mostly renaming types.

Demi-Marie committed:
Reading from an unwritten stream must fail

Since streams are opened lazily in QUIC, reading from a stream that has
never been written to cannot possibly succeed.  Return an error in this
case.

Demi-Marie committed:
Refactor connection code

- Use the existing Pending machinery to handle packet transmission
- Remove unnecessary looping in ‘ConnectionDriver::poll’ and
  ‘Muxer::drive_timer’

Demi-Marie committed:
Fix clippy lints and test endpoint shutdown

The code is now clippy-clean, and endpoint shutdown is tested properly.

Demi-Marie committed:
Only forbid warnings when testing

Future releases of Rust or Clippy may add new warnings, so a blanket
`#[forbid(warnings, clippy::all)]` is a bad idea.  However, the test
suite is only ran by developers and CI, so
`#[forbid(warnings, clippy::all)]` is fine there.

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Add Cargo metadata

Demi-Marie committed:
Wake up the endpoint driver when needed

When `Endpoint` or `EndpointRef` is dropped, the endpoint driver needs
to be woken up.  Make sure that happens.

tomaka reviewed (changes requested):
Here's a preliminary review that scratches the surface.
I'm going to dive in the logic in the next days.

Demi-Marie committed:
Respond to code review and fix tests

In addition to problems pointed out by @tomaka, three of the tests were
broken.  A bug in the tests meant they passed anyway.

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Move EndpointMessage to endpoint.rs

where it logically belongs.

Demi-Marie committed:
Reformat

Demi-Marie committed:
Move the configuration code into the endpoint

Demi-Marie committed:
Fix comment

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Fix broken links in rustdoc

tomaka reviewed (changes requested):
Another pass.
I haven't really started dwelling into `endpoint.rs` and `connection.rs` because they're really opaque.

Demi-Marie committed:
Apply suggestions from code review

Co-Authored-By: Pierre Krieger <pierre.krieger1708@gmail.com>

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Refactor as per @tomaka’s suggestion

Demi-Marie committed:
Implement Transport for Endpoint

Demi-Marie committed:
QUIC is not QUIC/UDP/IP

Co-Authored-By: Pierre Krieger <pierre.krieger1708@gmail.com>

Demi-Marie committed:
Avoid &Arc<Self>, as it is unstable.

Demi-Marie committed:
Merge branch 'demi-quic-stable-futures' of github.com:DemiMarie-parity/rust-libp2p into demi-quic-stable-futures

Demi-Marie committed:
Avoid references from connection to endpoint

This also adds better tests and makes other improvements.

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Avoid use of deprecated elided lifetimes

Demi-Marie committed:
Add a bunch of documentation

Demi-Marie committed:
Fix broken links in documentation

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Also reformat and fix compilation and tests

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Bump dependencies of libp2p-quic (only)

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Be more selective about enabling lints

Instead of a blunderbuss `#[forbid(warnings)]`, this denys certain lints
that are likely not to cause further problems in the future, including
some that are `Allow` by default. This required fixing the code in some
places, mostly replacing `pub` with `pub(crate)`.

Demi-Marie committed:
Merge the connection hashmaps and split them into a separate module

Demi-Marie committed:
`Muxer` now uses a wrapper around `Connection`

This avoids the connection needing to talk to the endpoint directly.

Demi-Marie committed:
Consolidate error messages

Demi-Marie committed:
Tests should not require `Outbound` to implement `Stream`

Demi-Marie committed:
Outbound should not implement Future

Demi-Marie committed:
Move timer and last timeout to ConnectionDriver

The Muxer never uses them.

Demi-Marie committed:
Enhanced logging

Demi-Marie committed:
Fix shutdown and clean up the code

This avoids a reference count underflow when closing connections. It
also cleans up the code and documents some internal APIs.

Demi-Marie committed:
Fix doc comments and bump dependencies

We are stuck with relying on a (temporary) async-tls fork while we wait
on https://github.com/async-std/async-tls/19 to be merged.

Demi-Marie committed:
Fix compilation error in websockets

Demi-Marie committed:
Adapt to changed rustls API

‘ClientCertVerifier::client_auth_root_subjects()’ previously returned
a ‘Vec’, but now returns an ‘Option<Vec>’.  As a result,
‘Default::default()’ is now ‘None’, not ‘vec![]’.  This causes the
connection to be aborted instead of allowing any subject.

Fix this by explicitly returning ‘Some(vec![])’.

There is still a bug: if the TLS handshake fails, the code will panic
when it fails to obtain the nonexistent peer certificate.  This will be
fixed in the next commit.

Demi-Marie committed:
Fix error handling

A connection does not need to be drained to be closed.  This could cause
a panic if the TLS handshake failed: the code would expect there to
be a peer certificate, but the handshake failure meant there wasn’t any.

Demi-Marie committed:
Use the public type alias for ‘DistinguishedNames’

instead of the internal one.

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Fix intra-rustdoc links

Demi-Marie committed:
Use a temporary fork of webpki

Demi-Marie committed:
Adapt to changed extension handling API in webpki

Demi-Marie committed:
Delete unneeded code and refactor

This cleans up the code and makes it easier to understand.

Demi-Marie committed:
Simplify code even more

This does additional refactoring to make the code smaller and easier to
understand.

Demi-Marie committed:
Clean up certificate verification

Demi-Marie committed:
Refactor TLS and connection code

This prepares for the TLS code being moved into a separate crate.

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Move libp2p-tls to its own crate

Demi-Marie committed:
Move stream_map and stream up a directory

Demi-Marie committed:
Move task wakeup into stream_map.rs

The code responsible for waking up tasks had previously been intermixed
with the code that adapts quinn-proto’s own API to that of libp2p. This
separates them, which makes the code significantly clearer and easier to
understand.

Demi-Marie committed:
Inline functions used only once

Demi-Marie committed:
Remove git submodule

Demi-Marie committed:
More cleanups

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Fix panic when connection closed too soon

The connection may be lost before the handshake is complete.  If this
happens, the connection will not be handshaking, but there will not be
any certificates available.  Return an error in this case instead of
panicing.

Demi-Marie committed:
Be more rigorous about send stream counts

They must be decremented as soon as we know we will never write on that
stream again, and they must not be decremented twice for the same
stream.

Demi-Marie committed:
Use quinn-proto’s built-in send stream count

instead of trying to count send streams manually.

Furthermore, stop using a fork of async-tls that has already been
merged.

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Use a patch instead of a git dependency

Demi-Marie committed:
Apply suggestions from Max Inden

Demi-Marie committed:
Cleanup docs and remove a callback

Thank you to Max Inden for the suggestions.

Demi-Marie committed:
Add a span for each test run

Demi-Marie committed:
Switch to a working quinn-proto branch

Also simplify error handling by panicing on certain impossible
conditions.

Demi-Marie committed:
Move the multiaddr to the configuration

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Switch back to quinn-proto master

Demi-Marie committed:
Fix compilation of tests

Demi-Marie committed:
Implement alternate X.509 verifier

Demi-Marie committed:
Fix doc tests

Demi-Marie committed:
protocols/tls needs dangerous_configuration

Demi-Marie committed:
Refactor certificate verification code

Demi-Marie committed:
Revert "Refactor certificate verification code"

This reverts commit fc39b7b1ebb47ac3a918b195791746935353605b.

Demi-Marie committed:
Distinguished unsupported from mismatching signature algorithms

Demi-Marie committed:
Replace the old webpki-based verifier

Demi-Marie committed:
Minor cleanups

Demi-Marie committed:
Remove use of webpki in certificate verification

Also add tracing

Demi-Marie committed:
Use a working quinn-proto version

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Remove unneeded feature in yasna

Demi-Marie committed:
Packets cannot send themselves

Demi-Marie committed:
Send an empty distinguished name

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Make tracing optional

Demi-Marie committed:
Update description of libp2p-tls

Demi-Marie committed:
libp2p-tls → libp2p-x509

Demi-Marie committed:
Fix silly macro error

Demi-Marie committed:
Bump async-tls

Demi-Marie committed:
Merge branch 'better-async-tls' into demi-quic-stable-futures

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Fix silly typos in Cargo.toml

Demi-Marie committed:
Avoid depending on pem

It’s a default feature of rcgen, but libp2p doesn’t need it.

Demi-Marie committed:
Implement certificate signature verification manually

Demi-Marie committed:
Switch to x509-signature crate

Demi-Marie committed:
Better error handling

Demi-Marie committed:
Switch to a released version of x509-signature

Demi-Marie committed:
Merge branch 'demi-quic-stable-futures' into external-x509

Demi-Marie committed:
Fix bogus unwrap

Demi-Marie committed:
Fix doc test

Demi-Marie committed:
Clean up copyright headers

Demi-Marie committed:
Apply suggestions from code review

Co-Authored-By: Toralf Wittner <tw@dtex.org>
Co-Authored-By: Max Inden <mail@max-inden.de>

Demi-Marie committed:
Merge branch 'demi-quic-stable-futures' of github.com:DemiMarie-parity/rust-libp2p into demi-quic-stable-futures

Demi-Marie committed:
Remove unwraps from certificate generation

Demi-Marie committed:
Clean up documentation and lints

Demi-Marie committed:
Sort dependencies

Demi-Marie committed:
Remove a potential panic and dead code

Demi-Marie committed:
Explain that QUIC resends lost packets

Demi-Marie committed:
We already ignore transmit I/O errors

so there is no need to handle them.

Demi-Marie committed:
Remove unused field

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Update to x509-signature 0.2.0

Demi-Marie committed:
Switch to the TLS patch that will be merged

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
err-derive ⇒ thiserror

for consistency with the rest of libp2p

Demi-Marie committed:
‘err-derive’ ⇒ ‘thiserror’ in libp2p-x509

This is also done for consistency.

Demi-Marie committed:
Switch to rustls master

It now supports user-defined signature verification.

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Bump rcgen

Demi-Marie committed:
Bump x509-signature dependency

This fixes a critical bug in the x509-signature crate, which fortunately
does not affect libp2p.  It also makes deprecation warnings no longer be
errors, as they can appear in minor releases of upstream dependencies.

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Upgrade to the newest x509-signature crate

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Update to quinn-proto and rustls master

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

tomaka committed:
Merge remote-tracking branch 'upstream/master' into quiccc-again

tomaka committed:
QUICk fix

tomaka committed:
Working prototype

tomaka committed:
Working well enough

tomaka committed:
Work

Demi-Marie committed:
Merge branch 'master' into demi-quic-stable-futures

Demi-Marie committed:
Trying to get the tests working

Demi-Marie committed:
Merge branch 'master' into quiccc-again

Demi-Marie committed:
Do not discard NewAddress messages

Demi-Marie committed:
Merge branch 'quiccc-again' into demi-quic-stable-futures

Demi-Marie committed:
Avoid hangs when running tests

Demi-Marie committed:
Wait until the handshake is complete

even if certificates are available sooner.

Also fix some warnings and rustdoc links.

Demi-Marie committed:
A handshake with no certificate is unfinished

Demi-Marie committed:
Re-enable several tests

Demi-Marie committed:
Fix some warnings in the tests

Demi-Marie committed:
Remove some commented-out test code

Demi-Marie committed:
Remove unused fields from the endpoint

Demi-Marie committed:
Avoid looping forever if no certificate is sent

Demi-Marie committed:
Working libp2p-quic (again)

This fixes the hang when running the test suite.

Demi-Marie committed:
Remove some disabled code
